src/components/action-popover/action-popover.component.tsx

```diff
@@ -70,249 +72,275 @@ export interface ActionPopoverProps extends MarginProps {
+        return "top-start";
+      }
 
-  const firstFocusableItem = findFirstFocusableItem(items);
+      if (placement === "bottom" && rightAlignMenu) {
+        return "bottom-start";
+      }
 
-  const lastFocusableItem = findLastFocusableItem(items);
+      return "bottom-end";
+    }, [placement, rightAlignMenu]);
 
-  invariant(
-    hasProperChildren,
-    `ActionPopover only accepts children of type \`${ActionPopoverItem.displayName}\`` +
-      ` and \`${ActionPopoverDivider.displayName}\`.`,
-  );
+    const setOpen = useCallback(
+      (value) => {
+        if (value && !isOpen) {
+          onOpen();
+        }
+        if (!value && isOpen) {
+          onClose();
+        }
+        setOpenState(value);
+      },
+      [isOpen, onOpen, onClose],
+    );
 
-  const mappedPlacement = useMemo(() => {
-    if (placement === "top" && !rightAlignMenu) {
-      return "top-end";
-    }
+    const focusButton = useCallback(() => {
+      const button = buttonRef.current?.querySelector<HTMLElement>(
+        "[data-element='action-popover-button']",
+      );
 
-    if (placement === "top" && rightAlignMenu) {
-      return "top-start";
-    }
+      button?.focus();
+    }, []);
 
-    if (placement === "bottom" && rightAlignMenu) {
-      return "bottom-start";
-    }
+    useImperativeHandle<ActionPopoverHandle, ActionPopoverHandle>(
+      ref,
+      () => ({
+        focusButton() {
+          focusButton();
+        },
+      }),
+      [focusButton],
+    );
 
-    return "bottom-end";
-  }, [placement, rightAlignMenu]);
+    const onButtonClick = useCallback(
+      (e) => {
+        e.stopPropagation();
+        const isOpening = !isOpen;
+        setFocusIndex(firstFocusableItem);
+        setOpen(isOpening);
+        if (!isOpening) {
+          // Closing the menu should focus the MenuButton
+          focusButton();
+        }
+      },
+      [isOpen, firstFocusableItem, setOpen, focusButton],
+    );
 
-  const setOpen = useCallback(
-    (value) => {
-      if (value && !isOpen) {
-        onOpen();
-      }
-      if (!value && isOpen) {
-        onClose();
-      }
-      setOpenState(value);
-    },
-    [isOpen, onOpen, onClose],
-  );
+    // Keyboard commands implemented as recommended by WAI-ARIA best practices
+    // https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-actions.html
+
+    const onButtonKeyDown = useCallback(
+      (e) => {
+        if (
+          Events.isSpaceKey(e) ||
+          Events.isDownKey(e) ||
+          Events.isEnterKey(e)
+        ) {
+          e.preventDefault();
+          e.stopPropagation();
+          setFocusIndex(firstFocusableItem);
+          setOpen(true);
+        } else if (Events.isUpKey(e)) {
+          e.preventDefault();
+          e.stopPropagation();
+          setFocusIndex(lastFocusableItem);
+          setOpen(true);
+        }
+      },
+      [firstFocusableItem, lastFocusableItem, setOpen],
+    );
```

Comment:
**suggestion (non-blocking):** Feels like this can be better written to remove some of the duplicated lines (just for neatness and readability, though - it's functionally fine). Something like this (via Co-Pilot):

```javascript
const onButtonKeyDown = useCallback(
  (e) => {
    if (
      Events.isSpaceKey(e) ||
      Events.isDownKey(e) ||
      Events.isEnterKey(e) ||
      Events.isUpKey(e)
    ) {
      e.preventDefault();
      e.stopPropagation();
      setFocusIndex(
        Events.isUpKey(e) ? lastFocusableItem : firstFocusableItem
      );
      setOpen(true);
    }
  },
  [firstFocusableItem, lastFocusableItem, setOpen],
);
```

Reply:
Also, a re-read indicates the original code was already there and this "change" was just because it moved, so feel free to completely ignore this 😄

Reply:
Will add this in for some cleanliness